src/psyclone/psyir/tools/definition_use_chains.py

```diff
@@ -0,0 +1,658 @@
+            for reference in region.walk((Reference, Call, CodeBlock, Return)):
+                # Store the position instead of computing it twice.
+                abs_pos = reference.abs_position
+                if abs_pos <= self._start_point or abs_pos >= self._stop_point:
+                    continue
+                if isinstance(reference, Return):
+                    # When we find a return statement any following statements
+                    # can be ignored so we can return.
+                    if defs_out is not None:
+                        self._defsout.append(defs_out)
+                    return
+                if isinstance(reference, CodeBlock):
+                    # CodeBlocks only find symbols, so we can only do as good
+                    # as checking the symbol - this means we can get false
+                    # positives for structure accesses inside CodeBlocks.
+                    if isinstance(reference._fp2_nodes[0], Goto_Stmt):
+                        raise InternalError("DefinitionUseChains can't handle "
+                                            "code containing GOTO statements.")
+                    # If we find an Exit or Cycle statement, we can't
+                    # reach further in this code region so we can return.
+                    if isinstance(reference._fp2_nodes[0], (Exit_Stmt,
+                                                            Cycle_Stmt)):
+                        if defs_out is not None:
+                            self._defsout.append(defs_out)
+                        return
+                    if (
+                        self._reference.symbol.name
+                        in reference.get_symbol_names()
+                    ):
+                        # Assume the worst for a CodeBlock and we count them
+                        # as killed and defsout and uses.
+                        if defs_out is not None:
+                            self._killed.append(defs_out)
+                        defs_out = reference
+                        continue
+                elif isinstance(reference, Call):
+                    # If its a local variable we can ignore it as we'll catch
+                    # the Reference later if its passed into the Call.
+                    if self._is_local:
+                        continue
+                    if isinstance(reference, IntrinsicCall):
+                        # IntrinsicCall can only do stuff to arguments, these
+                        # will be caught by Reference walk already.
+                        # Note that this assumption two symbols are not
+                        # aliases of each other.
+                        continue
+                    # For now just assume calls are bad if we have a non-local
+                    # variable and we count them as killed and defsout
+                    # and uses.
+#                    if defs_out is None:
+#                        self._uses.append(reference)
```

Comment:
Maybe it would be nicer to bring here the opposite of the first condition `if not self._is_local:` or `if not reference.is_automatic:` instead of having the code/comments separated.

Also can the commented out lines be removed?

Reply:
Yeah, the commented out lines can be removed.

I don't quite understand the comment otherwise, I prefer `if cond: continue` to skip here to have less nesting.